File: src/main/java/com/cinebook/service/PromotionService.java
package com.cinebook.service;

import com.cinebook.dto.request.CreatePromotionRequest;
import com.cinebook.dto.request.UpdatePromotionRequest;
import com.cinebook.dto.request.UpdatePromotionStatusRequest;
import com.cinebook.dto.response.PageResponse;
import com.cinebook.dto.response.PromotionResponse;
import com.cinebook.dto.response.ValidatePromotionResponse;
import com.cinebook.entity.Promotion;
import com.cinebook.enums.PromotionStatus;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;

public interface PromotionService {

    PageResponse<PromotionResponse> getAdminPromotions(PromotionStatus status, String keyword, Pageable pageable);

    PromotionResponse getPromotionDetail(String id);

    PromotionResponse createPromotion(CreatePromotionRequest request);

    PromotionResponse updatePromotion(String id, UpdatePromotionRequest request);

    PromotionResponse updatePromotionStatus(String id, UpdatePromotionStatusRequest request);

    ValidatePromotionResponse validatePromotionCode(String code, BigDecimal grossAmount);

    BigDecimal calculateDiscount(Promotion promotion, BigDecimal grossAmount);
}

File: frontend/src/services/promotion.service.ts
import apiClient from './api'
import type { PageResponse } from '@/types/api.types'
import type {
  PromotionResponse,
  ValidatePromotionResponse,
  CreatePromotionRequest,
  UpdatePromotionRequest,
  UpdatePromotionStatusRequest,
  PromotionStatus,
} from '@/types/promotion.types'

export const promotionService = {
  // Public / Customer validation preview
  async getPublicPromotions(params?: {
    page?: number
    size?: number
    sort?: string
  }): Promise<PageResponse<PromotionResponse>> {
    const response = await apiClient.get<PageResponse<PromotionResponse>>('/api/v1/promotions', {
      params,
    })
    return response.data
  },

  async validatePromotionCode(
    code: string,
    grossAmount: number
  ): Promise<ValidatePromotionResponse> {
    const response = await apiClient.get<ValidatePromotionResponse>('/api/v1/promotions/validate', {
      params: { code, grossAmount },
    })
    return response.data
  },

  // Admin Management Endpoints
  async getAdminPromotions(params?: {
    status?: PromotionStatus
    q?: string
    page?: number
    size?: number
    sort?: string
  }): Promise<PageResponse<PromotionResponse>> {
    const response = await apiClient.get<PageResponse<PromotionResponse>>('/api/v1/admin/promotions', {
      params,
    })
    return response.data
  },

  async getPromotionDetail(id: string): Promise<PromotionResponse> {
    const response = await apiClient.get<PromotionResponse>(`/api/v1/admin/promotions/${id}`)
    return response.data
  },

  async createPromotion(payload: CreatePromotionRequest): Promise<PromotionResponse> {
    const response = await apiClient.post<PromotionResponse>('/api/v1/admin/promotions', payload)
    return response.data
  },

  async updatePromotion(id: string, payload: UpdatePromotionRequest): Promise<PromotionResponse> {
    const response = await apiClient.put<PromotionResponse>(`/api/v1/admin/promotions/${id}`, payload)
    return response.data
  },

  async updatePromotionStatus(
    id: string,
    payload: UpdatePromotionStatusRequest
  ): Promise<PromotionResponse> {
    const response = await apiClient.patch<PromotionResponse>(
      `/api/v1/admin/promotions/${id}/status`,
      payload
    )
    return response.data
  },
}

export default promotionService

